docs/porting-guide.rst

```diff
@@ -2055,6 +2055,28 @@ power down state where as it could be either power down, retention or run state
 for the higher power domain levels depending on the result of state
 coordination. The generic code expects the handler to succeed.
 
+plat\_psci\_ops.pwr\_domain\_suspend\_\early() [optional]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Perform the platform specific actions to prepare to suspend the calling
+CPU and its higher parent power domain levels as indicated by the
+``target_state`` (first argument). It is called by the PSCI ``CPU_SUSPEND``
+API implementation.
+
+The ``target_state`` has a similar meaning as described in
+the ``pwr_domain_off()`` operation. It encodes the platform coordinated
+target local power states for the CPU power domain and its parent
+power domain levels. The handler needs to perform power management operation
+corresponding to the local state at each power level. The generic code
+expects the handler to succeed.
+
+This handler gets called with the data cache enabled to allow platforms
+to perform necessary actions before the CPU actually enters the suspend
+state.
+
+This platform handler is optional and can theoretically replace the
+``pwr_domain_suspend`` handler, when implemented.
+
```

Comment:
Rather than copying and pasting the text from `pwr_domain_suspend()`, I think it would be clearer to merge these together in the same section. The text needs to make it clear that:

*   One or both of these functions must be present
*   One of the functions is called before CPU specific power down operations are performed, and one after.
*   On `HW_ASSISTED_COHERENCY = 0` platforms, one is called with data caches enabled, the other with data caches disabled. On `HW_ASSISTED_COHERENCY = 1` platforms, the data caches are enabled throughout.